docs/reference/programming_model.rst

```diff
@@ -0,0 +1,259 @@
+algorithms, but most importantly a matching threading model to think in terms
+of. It relaxes some of the restrictions of the :ref:`inherent_thread_model`
+imposed by the strict 1:1 mapping of architectural details to the programming
+model.
+
+The rich set of APIs introduced by Cooperative Groups allow the programmer
+to define their own groups based on run-time predicates, but a set of implicit
+groups manifest based on kernel launch parameters.
+
+The thread hiearchy abstraction of Cooperative Groups manifest as depicted in
+:numref:`coop_thread_hierarchy`.
+
+.. _coop_thread_hierarchy:
+
+.. figure:: ../data/reference/programming_model/thread_hierarchy_coop.svg
+  :alt: Diagram depicting nested rectangles of varying color. The outermost one
+        titled "Grid", inside sets of different sized rectangles layered on
+        oneanother titled "Block". Each "Block" containing sets of uniform
+        rectangles layered on oneanother titled "Warp". Each of the "Warp"
+        titled rectangles filled with downward pointing arrows inside.
+
+  Cooperative group thread hierarchy.
+
+* Multi Grid is an abstraction of potentially multiple simultaneous launches of
+  the same kernel over multiple devices. Grids inside a multi device kernel
+  launch need not be of uniform size, thus allowing taking into account
+  different device capabilities and preferences.
+
+  .. deprecated:: 5.0
+
+* Same as the :ref:`inherent_thread_model` Grid entity. The ability to
+  synchronize over a grid requires the kernel to be launched using the
+  Cooperative Groups API.
+
+* The defining feature of a cluster or block cluster is that all threads in a
+  cluster will share a common set of distributed shared memory which they may
+  use to share data or synchronize with oneanother.
+
+* Same as the :ref:`inherent_thread_model` Block entity.
+
+.. note::
+
+  Explicit warp-level thread handling is absent from the Cooperative Groups API.
+  In order to exploit the known hardware SIMD width on which built-in
+  functionality translates to simpler logic, one may use the group partitioning
+  part of the API, typically but not necessarily ``tiled_partition``.
+
+Memory Model
+===============================================================================
+
+The hierarchy of threads introduced by :ref:`inherent_thread_model` is induced
+by the memory subsystem of GPUs. :numref:`memory_hierarchy` summarizes that memory namespaces and
+how they relate to the various levels of the threading model.
+
+.. _memory_hierarchy:
+
+.. figure:: ../data/reference/programming_model/memory_hierarchy.svg
+  :alt: Diagram depicting nested rectangles of varying color. The outermost one
+        titled "Grid", inside on the upper half a rectangle titled "Cluster".
+        Inside it are two identical rectangles titled "Block", inside them are
+        ones titled "Local" with multiple "Warp" titled rectangles. Blocks have
```

Comment:
Same criticism as earlier, this being written in passive makes it harder to understand.
Actually, this one has a good example of how to make it less ambiguous - by adding "it" after "inside", it already makes a huge difference. Compare:
```suggestion
  :alt: Diagram depicting nested rectangles of varying color. The outermost one
        titled "Grid", inside (of it) on the upper half (is) a rectangle titled "Cluster".
        Inside it are two identical rectangles titled "Block", inside them are
        ones titled "Local" with multiple "Warp" titled rectangles. Blocks have
```

Reply:
Fixed